File: 016_MaxConsecutiveOnesIII.py
"""
Topic: Sliding Window / Two Pointers
Max Consecutive Ones III: longest subarray with at most k zeros
Time Complexity: O(n)
Space Complexity: O(1)
"""

def longest_ones(nums, k):
    left = 0
    zeros = 0
    best = 0
    for right, v in enumerate(nums):
        if v == 0:
            zeros += 1
        while zeros > k:
            if nums[left] == 0:
                zeros -= 1
            left += 1
        best = max(best, right - left + 1)
    return best


if __name__ == '__main__':
    print(longest_ones([1,1,1,0,0,0,1,1,1,1,0], 2))  # 6
    print(longest_ones([0,0,1,1,0,0,1,1,1,0,1,1], 3))  # 6

File: 016_MaxConsecutiveOnesIII.cpp
// Topic: Sliding Window / Two pointers
// Max Consecutive Ones III: Find the longest subarray of 1s, allowing you to flip at most k zeros.
// Time Complexity: O(n)
// Space Complexity: O(1)

#include <bits/stdc++.h>
using namespace std;

int longestOnes(vector<int>& nums, int k)
{
    int left = 0, zeros = 0, best = 0;
    for (int right = 0; right < (int)nums.size(); ++right)
    {
        if (nums[right] == 0)
            ++zeros;
        while (zeros > k)
        {
            if (nums[left] == 0)
                --zeros;
            ++left;
        }
        best = max(best, right - left + 1);
    }
    return best;
}

int main()
{
    vector<int> a1 = {1, 1, 1, 0, 0, 0, 1, 1, 1, 1, 0};
    cout << longestOnes(a1, 2) << '\n';  // 6
    vector<int> a2 = {0, 0, 1, 1, 0, 0, 1, 1, 1, 0, 1, 1, 0, 0, 0, 1, 1, 1, 1};
    cout << longestOnes(a2, 3) << '\n';
    vector<int> a3 = {1, 0, 0, 1, 1, 0, 1};
    cout << longestOnes(a3, 1) << '\n';
    return 0;
}
